sympy/sets/conditionset.py

```diff
@@ -34,6 +34,8 @@ def __new__(cls, sym, condition, base_set):
             return S.EmptySet
         if condition == S.true:
             return base_set
+        if getattr(base_set, "is_FiniteSet", False) and base_set.is_FiniteSet:
+            return base_set.intersect(condition.as_set())
```

Comment:
If the condition cannot be converted to a set then another approach is needed and is perhaps the preferred anyway,

```python
>>> ConditionSet(x, Eq(x%2,0), FiniteSet(1,2,3,4))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "sympy\sets\conditionset.py", line 38, in __new__
    return base_set.intersect(condition.as_set())
  File "sympy\core\relational.py", line 225, in as_set
    return solve_univariate_inequality(self, sym, relational=False)
  File "sympy\solvers\inequalities.py", line 422, in solve_univariate_inequality

    solns = solve(e, gen, check=False)
  File "sympy\solvers\solvers.py", line 1053, in solve
    solution = _solve(f[0], *symbols, **flags)
  File "sympy\solvers\solvers.py", line 1619, in _solve
    raise NotImplementedError('\n'.join([msg, not_impl_msg % f]))
NotImplementedError:
No algorithms are implemented to solve equation Mod(_Dummy_49, 2)
```

I would recommend something based on the following:

```python
>>> def resolve(x, cond, fs):
...   from sympy.utilities.iterables import sift
...   from sympy.core.logic import fuzzy_bool
...   sifted = sift(fs, lambda _: fuzzy_bool(cond.subs(x, _)))
...   return Union(FS(*sifted[True]), ConditionSet(x, cond, FS(*sifted[None])))
...
>>> from sympy import *
>>> from sympy.abc import x, y
>>> FS = FiniteSet
>>> t=FS(1, 4)
>>> f=FS(2)
>>> u=FS(y)
>>> resolve(x, Eq(x%3,1), t+f+u)
{1, 4} U ConditionSet(x, Eq(Mod(x, 3), 1), {y})
```

One of the advantages of this, too, is that it leaves the ConditionSet in the same form it was received.